src/R/Core/Impl/AST/Extensions/ScopeExtensions.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Common.Core;
+using Microsoft.R.Core.AST.Arguments;
+using Microsoft.R.Core.AST.DataTypes;
+using Microsoft.R.Core.AST.Functions.Definitions;
+using Microsoft.R.Core.AST.Scopes;
+using Microsoft.R.Core.AST.Scopes.Definitions;
+using Microsoft.R.Core.AST.Statements.Definitions;
+using Microsoft.R.Core.AST.Variables;
+
+namespace Microsoft.R.Core.AST {
+    public static class ScopeExtensions {
+        /// <summary>
+        /// Enumerates scopes from the current up to the global scope.
+        /// </summary>
+        public static IEnumerable<IScope> EnumerateTowardsGlobal(this IScope scope) {
+            yield return scope;
+            while (!(scope is GlobalScope)) {
+                scope = scope.GetEnclosingScope();
+                yield return scope;
+            }
+        }
+
+        /// <summary>
+        /// Enumerates definitions of variables applicable to the given scope.
+        /// Traverses scopes from the provided scope up and enumerates
+        /// variables that appear at the left side of the left-hand assignment
+        /// operator or at the right side of the right-hand assignmend operator.
+        /// Includes both regular variables as well as function definitions.
+        /// Includes assignments that appear in the file up to the specified
+        /// position except in the global scope it enumerates all assignments.
```

Comment:
Right, so it does the right thing (or at least the better thing, given our constraints) in the global scope. What I'm suggesting is that the same logic - list everything, even if it's assigned after the current line - should be applied to all other scopes, because the same case where code in the middle can refer to code at the end, also exists there.